rspace/src/main/scala/coop/rchain/rspace/nextgenrspace/history/HistoryRepositoryImpl.scala

```diff
@@ -0,0 +1,224 @@
+package coop.rchain.rspace.nextgenrspace.history
+
+import cats.Applicative
+import cats.effect.Sync
+import cats.implicits._
+import coop.rchain.rspace.{
+  internal,
+  Blake2b256Hash,
+  DeleteContinuations,
+  DeleteData,
+  DeleteJoins,
+  HotStoreAction,
+  InsertContinuations,
+  InsertData,
+  InsertJoins,
+  Serialize,
+  StableHashProvider
+}
+import coop.rchain.rspace.internal._
+import scodec.{Attempt, Codec, DecodeResult, Decoder, Encoder, Err, SizeBound}
+import scodec.bits.BitVector
+import scodec.codecs.int32
+import HistoryRepositoryImpl.codecSeq
+
+final case class HistoryRepositoryImpl[F[_]: Sync, C, P, A, K](
+    history: History[F],
+    leafStore: ColdStore[F]
+)(implicit codecC: Codec[C], codecP: Codec[P], codecA: Codec[A], codecK: Codec[K])
+    extends HistoryRepository[F, C, P, A, K] {
+
+  def codecSeqC: Codec[Seq[C]] = codecSeq(codecC)
+
+  implicit def codecSeqSeqC: Codec[Seq[Seq[C]]] =
+    codecSeq(codecSeqC)
+
+  implicit def codecSeqDatum: Codec[Seq[internal.Datum[A]]] =
+    codecSeq(internal.codecDatum(codecA))
+
+  implicit def codecSeqWaitingContinuation: Codec[Seq[internal.WaitingContinuation[P, K]]] =
+    codecSeq(internal.codecWaitingContinuation(codecP, codecK))
+
+  val joinSuffixBits = BitVector.apply("-joins".getBytes)
+
+  private def hashChannels(channels: Seq[C]): Blake2b256Hash =
+    StableHashProvider.hash(channels)(Serialize.fromCodec(codecC))
+
+  private def hashChannel(channel: C): Blake2b256Hash =
+    StableHashProvider.hash(channel)(Serialize.fromCodec(codecC))
+
+  private def fetchData(key: Blake2b256Hash): F[Option[PersistedData]] =
+    history.findPath(key.bytes.toSeq.toList).flatMap {
+      case (trie, _) =>
+        trie match {
+          case Leaf(dataHash) => leafStore.get(dataHash)
+          case EmptyTrie      => Applicative[F].pure(None)
+          case _ =>
+            Sync[F].raiseError(new RuntimeException(s"unexpected data at key $key, data: $trie"))
+
+        }
+    }
+
+  def hashJoinsChannel(channel: C): Blake2b256Hash = {
+    val channelBits     = codecC.encode(channel).get
+    val joinChannelBits = channelBits ++ joinSuffixBits
+    Blake2b256Hash.create(joinChannelBits.toByteVector)
+  }
+
+  override def getJoins(channel: C): F[Seq[Seq[C]]] = {
+    val key = hashJoinsChannel(channel)
+    for {
+      storedLeaf <- fetchData(key)
+      result = storedLeaf match {
+        case Some(JoinsLeaf(bytes)) =>
+          codecSeqSeqC.decode(bytes.bits).get.value
+        case _ => Seq.empty
```

Comment:
Wouldn't raising an error be a better idea?

Reply:
an error would mean that a join has to exist. it would be impossible to test if a join exist without catching errors. not sure that's a good design choice. do you think it's useful?

Reply:
Right. What if due to a bug a different type of leaf was stored there?